File: Search/BinarySearch.py
#using recursion
def search(arr, start, end, to_search):
	if start>end:
		return -1

	mid = int(start + ((end-start)/2))	

	if arr[mid] == to_search:
		return mid
	elif to_search > arr[mid]:
		return search(arr, mid+1, end, to_search)
	else:
		return search(arr, start, mid-1, to_search)


#using itteration
def search(arr, to_search):
	start = 0
	end = len(arr) - 1

	while start<=end:
		mid = int(start + ((end-start)/2))

		if arr[mid] == to_search:
			return mid
		elif to_search < arr[mid]:
			end = mid -1
		else:
			start = mid+1

	return -1

#main functionality
arr = [8,9,7,5,2,10,7,1,6]
arr.sort()

#ans = search(arr, 0, len(arr)-1, 7) #end will the index of the last element
#function overloading is not allowed in python

ans2 = search(arr, 1)
print(ans2)

File: bhemu_jee/BinarySearch.java
class BinarySearch {
    public static void main(String args[]) {
        int arr [] = {5,7,8,11,13,16,20,22};
        boolean tf = obs(arr, 12);
        System.out.print(tf);
        

    }

    public static boolean search(int arr[], int toFind) {
        for (int a = 0; a < arr.length; a++) {
            int b = arr[a];
            if (b == toFind) {
                return true;
            }
        }
        return false;
    }
    
    public static boolean bs(int arr[], int toFind) {
        int mid = arr.length / 2; 
        if (toFind == arr[mid]) {
            return true;
        } else if (toFind < arr[mid]) {
            for (int a = mid; a >= 0; a--) {
                int value = arr[a];
                if (value == toFind) {
                    return true;
                }
            }
        } else {
            for (int a = mid; a < arr.length; a++) {
                int value = arr[a];
                if (value == toFind) {
                    return true;
                }
            }
        }
        return false;
    }
    public static boolean obs(int arr[], int toFind) {
        int start = 0;
        int end = arr.length-1;
        int mid = (start + end) / 2;
        while(start<end) {
            if (toFind == arr[mid]) {
                return true;
            } else if (toFind < arr[mid]) {
                end = mid-1;
            } else {
                start = mid + 1;
            }
            mid = (start + end) / 2;
        }
        return false;
    }
}